[BugFix] [sr-sqla] Fix DEFAULT compilation, type comparison, and column ordering

jaogoy opened the pull request:
## Why I'm doing:

Improvements and bug fixes for DEFAULT value compilation, type comparison, column ordering, and documentation.

## What I'm doing:

### 1. Process DEFAULT compilation to handle function and output string

Enhanced DEFAULT value compilation to properly handle function expressions and string outputs:

- Function defaults (e.g., `uuid()`, `NOW(6)`) are wrapped as `(func_expr)`
- DATETIME CURRENT_TIMESTAMP defaults are emitted verbatim
- Literal values are properly quoted with `quote`.

### 2. Add `get_col_spec` for StructuredType

Added `get_col_spec` method to `StructuredType` base class. Required by `sqlacodegen` for generating column specifications for structured types (ARRAY, MAP, STRUCT).

### 3. Ignore display_width when comparing Integer types

Fixed type comparison to ignore `display_width` for Integer types (TINYINT, SMALLINT, INTEGER, BIGINT, LARGEINT). Updated type compiler to ensure FLOAT/DOUBLE do not render display_width, as StarRocks doesn't support precision for these types.

### 4. Sort columns according to column_ordinal

Fixed column ordering in table reflection by sorting columns by `ORDINAL_POSITION` from information schema.

### 5. Update sqlacodegen option names in README.md

Updated README.md to use correct `sqlacodegen` option format: `include_dialect_options,keep_dialect_types`.

## Testing

- Unit tests for DEFAULT value rendering and type comparison
- Integration tests for column ordering
- All existing tests pass

## Impact

- **Breaking Changes**: None
- **Backward Compatibility**: All changes are backward compatible

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Improves default value compilation and structured type support, refines type/reflection comparisons (e.g., integer display width, BOOLEAN↔TINYINT(1)), sorts reflected columns by position, and updates docs/tests with a version bump.
> 
> - **Dialect/Compiler**:
>   - DEFAULT rendering: emit functions as `(expr)`, keep DATETIME `CURRENT_TIMESTAMP` verbatim, quote literals.
>   - Type compilation tweaks: map `VARCHAR(65533)`/unbounded `VARCHAR` to `STRING`, add `DOUBLE`, preserve `LARGEINT(display_width)`, ignore `FLOAT` precision.
>   - Identifier preparer and DDL compilers adjusted; create table key validation refined.
> - **Reflection/Parser**:
>   - Sort reflected columns by `ORDINAL_POSITION` and derive PK columns accordingly.
>   - Parse `TINYINT(1)` as `BOOLEAN` during reflection.
>   - Column defaults normalized as above; comments for VIEW/MV handled; CLOUD_NATIVE engine normalized to `OLAP`.
>   - Detect and cache system `run_mode`; filter invalid table properties per run mode.
> - **Alembic compare**:
>   - Type comparison ignores integer display_width; treats `BOOLEAN`≡`TINYINT(1)` and `STRING`≡`VARCHAR(65533)`.
>   - View/MV include-object checks made case-safe; property comparison uses purified properties.
> - **Data types**:
>   - Add `get_col_spec` and recursive `get_sub_item_types` for `ARRAY`/`MAP`/`STRUCT` to support codegen.
> - **Docs**:
>   - Fix `sqlacodegen` option names to `include_dialect_options,keep_dialect_types` and add guidance.
> - **Tests/Version**:
>   - Extensive unit/integration/system tests updated/added for defaults, types, reflection, Alembic; bump version to `1.3.2`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit cae7f6bfbe4700e8177a4daf859003e4703329d9. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

jaogoy committed:
Fix DEFAULT compilation, type comparison, and column ordering

## Summary

Improvements and bug fixes for DEFAULT value compilation, type
comparison, column ordering, and documentation.

## Changes

### 1. Process DEFAULT compilation to handle function and output string

Enhanced DEFAULT value compilation to properly handle function
expressions and string outputs:

- Function defaults (e.g., `uuid()`, `NOW(6)`) are wrapped as
`(func_expr)`
- DATETIME CURRENT_TIMESTAMP defaults are emitted verbatim
- Literal values are properly quoted

### 2. Add `get_col_spec` for StructuredType

Added `get_col_spec` method to `StructuredType` base class. Required by
`sqlacodegen` for generating column specifications for structured types
(ARRAY, MAP, STRUCT).

### 3. Ignore display_width when comparing Integer types

Fixed type comparison to ignore `display_width` for Integer types
(TINYINT, SMALLINT, INTEGER, BIGINT, LARGEINT). Updated type compiler to
ensure FLOAT/DOUBLE do not render display_width, as StarRocks doesn't
support precision for these types.

### 4. Sort columns according to column_ordinal

Fixed column ordering in table reflection by sorting columns by
`ORDINAL_POSITION` from information schema.

### 5. Update sqlacodegen option names in README.md

Updated README.md to use correct `sqlacodegen` option format:
`include_dialect_options,keep_dialect_types`.

## Testing

- Unit tests for DEFAULT value rendering and type comparison
- Integration tests for column ordering
- All existing tests pass

## Impact

- **Breaking Changes**: None
- **Backward Compatibility**: All changes are backward compatible

Signed-off-by: jaogoy <jaogoy@gmail.com>

jaogoy committed:
Fix DEFAULT reflection

1. Process DEFAULT "" (support empty string)
2. upper table_kind.

Signed-off-by: jaogoy <jaogoy@gmail.com>

jaogoy committed:
fix format

Signed-off-by: jaogoy <jaogoy@gmail.com>

jaogoy committed:
fix NPE

Signed-off-by: jaogoy <jaogoy@gmail.com>

jaogoy committed:
version +1

Signed-off-by: jaogoy <jaogoy@gmail.com>

jaogoy committed:
add docs info for using sqlacodegen: nullable and reorder

Signed-off-by: jaogoy <jaogoy@gmail.com>

rad-pat commented:
> @rad-pat Would you like to have a review?

@jaogoy, sure - I will pull and check SQLA Tests also run for me.

rad-pat reviewed (commented):
@jaogoy  - 1 minor comment. SQLAlchemy tests all run successfully, but your Starrocks specific tests, I get many failures which is presumably not expected. I also presume this is because of our setup of Starrocks in K8s using GCS - tests should probably be able to run on any system though?

jaogoy commented on contrib/starrocks-python-client/starrocks/dialect.py:
```diff
@@ -734,8 +732,10 @@ def get_column_specification(self, column: sa_schema.Column, **kw: Any) -> str:
             colspec.append(agg_info)
 
         # NULL or NOT NULL.
+        is_nullable_set = False
```
I'll remove it.
What I want to express is that if `NOT NULL` is not rendered, then it will render as `NULL`. But, it's not necessary because it't the ANSI-SQL behavior.

jaogoy committed:
[Bugfix] [sr-sqla] Fix reflection for shared-data mode: ENGINE and properties

## Summary

This PR fixes table reflection issues in shared-data mode by:

1. converting `CLOUD_NATIVE` engine to `OLAP` during reflection.
2. filtering out invalid table properties based on the system run mode.
3. processing default MV comment.

## Problem

In shared-data mode:

1. StarRocks uses `CLOUD_NATIVE` as the table engine: OLAP for show
create table, but CLOUD_NATIVE for information_schema.tables_config.
2. Some table properties (e.g., `storage_medium`,
`storage_cooldown_time`, `replicated_storage`) are invalid in
shared-data mode but are still being compared

## Solution

1. **Engine conversion**: Convert `CLOUD_NATIVE` to `OLAP` during table
reflection to maintain consistency
2. **Property filtering**: Introduce `InvalidTableProperty` class to
filter out architecture-specific invalid properties during comparison
3. **Test fixes**: Update test cases to handle both shared-data and
shared-nothing modes

## Changes

- **`starrocks/reflection.py`**: Convert `CLOUD_NATIVE` engine to `OLAP`
when parsing table options
- **`starrocks/common/params.py`**: Add `InvalidTableProperty` class to
filter invalid properties based on run mode
- **`starrocks/alembic/compare.py`**: Use property filtering when
comparing table properties
- **`starrocks/common/types.py`**: Add `CLOUD_NATIVE` to `TableEngine`
enum
- **Test files**: Update tests to check run mode and adjust assertions
accordingly

## Testing

- Updated integration tests to handle both shared-data and
shared-nothing modes
- Fixed system tests to wait for ALTER TABLE operations to complete
- All existing tests pass with the new changes

Signed-off-by: jaogoy <jaogoy@gmail.com>

alvin-celerdata commented:
@cursor review

rad-pat commented:
SQLAlchemy Tests ran succesfully 👍

jaogoy commented:
> > @jaogoy, changes look ok from inspection. Running tests and I still receive errors. The logs are attached, I hope it helps.
> > 
> > [test_errors.txt](https://github.com/user-attachments/files/23930783/test_errors.txt)
> 
> @rad-pat You need to run the integration/system test uisng `-n1`, otherwise test cases will interrupt each other.
> In the future, we can optimize these test cases to be cable to run parallelly.

@rad-pat so, is it ok to merge this pr?

rad-pat reviewed (approved):
@jaogoy, apologies - yes fine by me to merge it

alvin-celerdata commented:
@cursor review
